lib/wallet/src/Cardano/Wallet/Write/Tx/Balance.hs

```diff
@@ -293,16 +296,58 @@ instance Buildable (PartialTx era) where
         cardanoTxF :: Cardano.Tx era -> Builder
         cardanoTxF tx' = pretty $ pShow tx'
 
+-- validateUTxO :: UTxO -> TransactionLayer -> ...
+--
+data UTxOAssumptions = forall k ktype. UTxOAssumptions
+    { txLayer
+        :: TransactionLayer k ktype SealedTx
+        -- TODO: Replace with smaller and smaller parts of 'TransactionLayer'
+    , inputScriptLookup
+        :: Maybe ([(W.TxIn, W.TxOut)] -> [CA.Script KeyHash])
```

Comment:
Perhaps it would be worth making a type to encapsulate the input script lookup function?

Something like:
```hs
-- | Some documentation of our expectations about input script lookups.
--
newtype InputScriptLookup = InputScriptLookup
    {runInputScriptLookup :: [(W.TxIn, W.TxOut)] -> [CA.Script KeyHash]}
```

Then the `inputScriptLookup` field of `UTxOAssumptions` would become:
```patch
    , inputScriptLookup
-       :: Maybe ([(W.TxIn, W.TxOut)] -> [CA.Script KeyHash])
+       :: Maybe InputScriptLookup
```

Advantages:
- it provides a place to document our expectations about the behaviour of this function (which may not be obvious from the type signature alone).
- assuming this function is passed around internally, readers of the code can navigate to the definition of this type and read the documentation.

Reply:
Alternatively, could it be represented by a `Map`?

Then the `Nothing` case could just be represented by `mempty`. (Just thinking out loud.)

Reply:
I suspect it should be `Map ScriptHash -> Script`— or sure, maybe `Map something (CA.Script KeyHash)` — but I intentionally didn't want to set out to find the optimal type now, only to move it.

Reply:
Actually I think there's a point to using a function instead of a map: we're implementing it using a `isShared` function https://github.com/input-output-hk/cardano-wallet/blob/378d4b524b0690df1e9e31944e4561a46d3d80ad/lib/wallet/src/Cardano/Wallet.hs#L2406

Reply:
> I'm now opting to still 1) use a function 2) don't create a `newtype` wrapper (I suspect the need might be lessened if we were to have e.g. [this](https://input-output-rnd.slack.com/archives/C04T7AEDY4X/p1679422712991169?thread_ts=1678777670.296419&cid=C04T7AEDY4X)), but to simplify the type to `W.Address -> Script KeyHash` from `[(W.TxIn, W.TxOut)] -> [CA.Script KeyHash]`.

Okay, fair enough. We can always improve things later on, if we need to.

> to simplify the type to W.Address -> Script KeyHash from [(W.TxIn, W.TxOut)] -> [CA.Script KeyHash].
👍🏻